Initial build out

euwest opened the pull request:
implement a class which utilizes the Fetch API, adding some useful features:
- holds onto request configuration options passed during construction to send with all requests
- throws unsuccessful requests (http status code != 2XX) by default
- enables pre- and post-fetch hooks to be added, via `beforeRequest` and `afterResponse` method namespaces (either by extending the class or overriding the placeholders on specific instances)

for added error-handling convenience, during any request configuration step, you can add `okCodes` or a `test` function to be used on the response. `okCodes` should be an array of non-200 http status codes that are likewise acceptable, and shouldn't result in a thrown response. the `test` function takes the response as an argument and, if returns `false` for any reason will result in a thrown response.

euwest committed:
add default error-handling module, tests

will be used by stipulate class, allowing for easy configuring of rejection situations

euwest committed:
add coveralls, coverage badge to readme, travis success hook for reporting

euwest committed:
add lcov reporting to test script

euwest committed:
style corrections: semi-colons, multi-line objects

euwest committed:
build basic Stipulate class functionality and tests

prefix/postfix extension testing will come next, as a separate test file

euwest committed:
fix typo, link to es6 jshintrc

euwest committed:
add prefix/postfix invocations to send flow, add tests

send now runs the request through `.prefix` before fetching, and runs the response through `.postfix`
after to allow for extending either with desired transforms(like extracting json on all responses)

euwest committed:
single line after blocks
only performing one delete action, no need for multiline function def

euwest commented:
~~TODO:~~ use `lodash.pickby` to filter out headers with `null` or `''` values before passing them to the `Request` constructor. `deepdefaults` allows Stipulate#send to remove `baseOptions` headers by setting them to `null` or `''`, but the fetch polyfill [casts those values to strings](https://github.com/github/fetch/blob/4dbbfd0e09500ea969766edfd8289af1910ba135/fetch.js#L23-L38) and attaches them to the xhr if we don't remove them, which we don't want.

**edit:** done

euwest committed:
add tests for overriding headers, excluding null/empty headers

euwest committed:
add statusText as error msg when throwing, tests

LionOps commented:
I think we need a `.lintrc` file. Needs to be in master so Snitch will comment.

LionOps commented on .jshintrc:
```diff
@@ -1 +1 @@
-./node_modules/jsgreat/.jshintrc
```
Let's make this an `extends` instead of a symlink. I think we want to add `module: true` too. You'll need a second .jshintrc referencing lint-review's copy of jsgreat.

LionOps commented on src/index.js:
```diff
@@ -1 +1,32 @@
+class Stipulate {
+
+  constructor(options) {
+    this.baseOptions = options;
```
Should this be frozen or considered to be private? I don't think it should be manipulated after being created.

LionOps replied:
I think there should be a `baseUrl` too.

euwest replied:
I don't see a reason to prevent people from altering the base options post-creation if they want to. If they do so instead of creating a new instance I'm inclined to assume they understand the implications.

morrisallison replied:
I agree on the `baseUrl` option. It's something I definitely see us using a lot. If it's not present, I'd extend the class and add it in `beforeRequest` in every client.

euwest replied:
ok; what do you want it's behavior to be?

euwest replied:
After talking with @morrisallison about this, it's a definite possibility for the future but we're blocking other projects, and we need to get a prototype out the door. I'll open a feature request for it which y'all can chat on.

LionOps commented on src/index.js:
```diff
@@ -1 +1,32 @@
+    return request;
+  }
+
+  postfix(response) {
```
No `makeOkCheck` built in? How are you intending it be used?

LionOps replied:
Rename to `afterSend`?

euwest replied:
I agree that the name isn't particularly clear, but `afterSend` doesn't seem like an improvement. I get that it's saying "this happens after the request has been made" but what does that mean in terms of the response? Is the response involved? or is this just a hook to be executed once you've sent the request?

alternatives I've thought about:
- `onResponse`
- `whenReturned`
- `onceReceived`

I'm not particularly fond of any of those though

euwest replied:
and yes, `makeOkCheck` will be in here by default. thanks for reminding me

LionOps replied:
> I agree that the name isn't particularly clear, but afterSend doesn't seem like an improvement.

I semi-agree. It's not uncommon to have `before<Action>`, `on<Action>`, and `after<Action>`, and the action here is `send`. I'm iffy with `onResponse` because are you talking about receiving the response from the server or fetch's Response? `withRequest` and `withResponse`?

euwest replied:
> are you talking about receiving the response from the server or fetch's Response?

Can you clarify the distinction you're calling attention to?

LionOps replied:
I'm just saying that `onResponse` is potentially as confusing due to request/response existing outside of fetch. Although, I guess since this is fetch-based, it doesn't matter at all.

euwest replied:
I'm still not following; what do you mean by existing outside of fetch? My train of thought is:
- stipulate makes requests, and receives responses(doesn't matter whether you're talking about the classes specific to fetch or the concepts in general, same holds true)
- the "before" function is called before sending off a request(again, both are relevant) and gets passed the Request instance(fetch class)
- the "after" function is called after receiving a response(both relevant) and gets passed the Response instance(fetch class)

the upside i see to `onResponse` is that likewise it can be interpreted as either "on response" or "on Response" and both are accurate. 

perhaps `afterResponse` makes even more sense, as we have already received a response, and are performing an action after that fact. that would also give us nice symmetry if we used `beforeRequest` as well. 

Thoughts?

LionOps replied:
👍

LionOps commented on src/index.js:
```diff
@@ -1 +1,32 @@
+    this.baseOptions = options;
+  }
+
+  prefix(request) {
```
rename to `beforeSend`?

euwest replied:
potentially, but I'd like the pre and post methods to have name symmetry if possible, and `afterSend` is problematic as mentioned above. gonna return to this when we figure that one out

LionOps commented on src/index.js:
```diff
@@ -1 +1,32 @@
+    return response;
+  }
+
+  send(url, options) {
```
How are you expecting this to be used? I think that while the prefix/postfix is a nice feature, this doesn't go far enough in DRYing a client's code.

Remember, the requests in a client have an expected signature of `request(options, query)`. `options` is this case are things needed by the client, such as an id for a specific resource, POST data, etc. It's unrelated to the `options` in fetch's Request.

With what you have now, the clients will have to parse the querystring onto the `url` rather than letting stipulate handle it, as well as attaching the `body` to the `options` if it's present.  I don't think the latter is too terrible, but it does need to do null/undefined checks. URL parsing will need to be DRY too.

I think it's worth considering a signature of `send(url, options, body, query)` that takes client knowledge and converts it into fetch knowledge; else, the clients are going to have to do the conversion too.

Take a look at yosbbejs and consider how the requests would be rewritten to use stipulate.

euwest replied:
The module isn't designed to create urls for you though. I'm not opposed to writing one that does, but I'm unconvinced it should live here. And the body of a request is definitely an `option`, so I don't really see why we should pull it out into its own parameter.

euwest committed:
add ok check to send method, rename pre/postfix methods, fix tests

euwest committed:
jshintrc update, lintreview configuration

euwest committed:
add url resolving, extract option building into sub module, update tests

add sub module to deal with resolving url strings and query objects, to allow for passing those things separately.
move option building into a sub module to compartmentalize, and allow isolated testing.
change beforeRequest hook to take in and return a url and options, and happen prior to request object being made.
having access to the Request pre-fetch proved less useful than a hook for modifying url and merged options.
update and extend tests

euwest committed:
initial draft of more complete README, forgotten dependency addition

euwest committed:
tests for build-options sub module

lock in option building behavior with unit test suite

euwest committed:
remove unused chai import from build-options test file

euwest committed:
test resolve-url sub module

LionOps commented on .lintrc-lr:
```diff
@@ -0,0 +1,4 @@
+{
```
Wrong file name?

euwest committed:
rename lintrc-lr to actual: jshintrc-lr

snitch commented on test/index.js:
```diff
@@ -1,6 +1,152 @@
+      };
+      const stipulate = new Stipulate();
+
+      expect(stipulate.beforeRequest(url, options)).to.deep.equal([url, options]);
```
Line is too long.

snitch commented on test/pre-post-extension.js:
```diff
@@ -0,0 +1,54 @@
+  describe('with custom beforeRequest method', () => {
+    before(() => {
+      global.fetch = (url, options) => {
+        if(options.foo === 'bar') return Promise.resolve({ ok: true });
```
Expected '{' and instead saw 'return'.

snitch commented on src/index.js:
```diff
@@ -1 +1,34 @@
+      .then(checkOk)
+      .then(this.afterResponse);
+  }
+}
```
Missing semicolon.

snitch commented on src/make-ok-check.js:
```diff
@@ -0,0 +1,16 @@
+    error.response = response;
+
+    throw error;
+  }
```
Missing semicolon.

snitch commented on src/make-ok-check.js:
```diff
@@ -0,0 +1,16 @@
+
+    const codeOk = okCodes.includes(response.status);
+
+    if(!test && (response.ok || codeOk)) return response;
```
Expected '{' and instead saw 'return'.

snitch commented on src/make-ok-check.js:
```diff
@@ -0,0 +1,16 @@
+  }
+}
+
+export default makeOkCheck
```
Missing semicolon.

snitch commented on src/make-ok-check.js:
```diff
@@ -0,0 +1,16 @@
+function makeOkCheck({ okCodes = [], test } = {}) {
+  return (response) => {
+    if(test && test(response)) return response;
```
Expected '{' and instead saw 'return'.

snitch commented on test/resolve-url.js:
```diff
@@ -0,0 +1,39 @@
+      fizz: 'buzz'
+    };
+
+    expect(resolveUrl('http://some.domain/foo')).to.equal('http://some.domain/foo');
```
Line is too long.

snitch commented on test/build-options.js:
```diff
@@ -0,0 +1,79 @@
+    expect(buildOptions(a, b)).to.deep.equal(expected);
+  });
+
+  it('doesnt merge `okCodes`. first option set overrides second if both present', () => {
```
Line is too long.

euwest committed:
linting error fixes

snitch commented on src/index.js:
```diff
@@ -1 +1,34 @@
+
+    return fetch(...requestArguments)
+      .then(checkOk)
+      .then(this.afterResponse);
```
Unnecessary semicolon.

LionOps replied:
`bind` or fat arrow.

euwest replied:
Are you saying `this.afterResponse` should have a bound context? that's easily achieved when extending the class or adding the override method to the instance, so I'm not sure I'm on board with forcing it in all cases.

LionOps replied:
Eh.... I guess it's not needed, but the context of `afterResponse`  is the promise unless it's binded. It strikes me as wrong to be calling `this.x` when the context of the function won't be executed in the context of `this`.

Reason: These functions are designed to be extended, and since it's a class, it's reasonable to assume that the extended functions are executed in the context of the object and that `this.x` is available from within the function.

Say I want to attach a new property to the class, `ExtStip.foo` and I want to do something with it in the `afterResponse` hook. I can't, unless I also extend the `send` method or do some unintuitive bindings in the `afterResponse`. 

By design, `send` should never be extended. You're giving functions that are executed before/after send does it's thing, and if someone does need their own version of `send`, they probably don't want Stipulate to begin with.

Maybe expecting the class to be extended is also the wrong approach here. Maybe we should be providing hooks instead.

``` js
...
  _afterResponse(response) {
     return (response) => {
        if (this.afterResponse && typeof this.afterResponse === 'Function') {
          return this.afterResponse(response);
        }

        return response;
     }();
  }
...
```

NOTE: Above is an example of the unintuitive binding I was talking about.

Actually, I think we should do this approach either way. `beforeRequest` and `afterResponse` can be set during an extension or on the fly (I might not want to create a subclass just to access hooks), and the decision is up to the developer.

euwest replied:
Ah, damn. Looks like:

``` js
class Foo {
  afterResponse = (response) => {
    ...
  }
}
```

is stage 1. Alright, I'll bind it in `send`.

> Maybe expecting the class to be extended is also the wrong approach here. Maybe we should be providing hooks instead.

The idea is to allow for both, but I'm beginning to think we have different working definitions of "providing hooks". Would you elaborate?

LionOps replied:
Yeah.

I would not call `beforeResponse` and `afterResponse` directly in `send` since `send` is the core of Stipulate and shouldn't be extended/overwritten (If you do extend it, it'd have to be completely rewritten). Instead, I would do `_beforeResponse` and `_afterResponse` that check for the existence of `beforeResponse` and `afterResponse` respectively, that they are functions, and call them if so.

Doing this has two advantages:

1) The hooks are isolated from `send`, and you don't need a default hook that does a passthrough.
2) You wouldn't need to write tests for defaults that don't do anything. Instead, you'd write tests that assert the hooks were called if they were set. This is good because the tests actually match the intended use.

euwest replied:
> The hooks are isolated from send, and you don't need a default hook that does a passthrough.

your proposed `_before*` and `_after*` functions are passthroughs if `before*` and `after*` don't exist, are they not? I think you're going to have to show me a code example of what you're talking about.

> You wouldn't need to write tests for defaults that don't do anything. Instead, you'd write tests that assert the hooks were called if they were set.

it really sounds like you're proposing a middle man between `send` and the hooks that does the exact same thing that's currently happening.

LionOps replied:
> it really sounds like you're proposing a middle man between send and the hooks that does the exact same thing that's currently happening.

It's a subtle difference for stronger code.

``` js
class ThingWithHooks {
  _beforeHook(value) {
    if (this.before && typeof this.before === 'function') {
      return this.before(value);
    }

    return value;
  }

  _afterHook(value) {
    if (this.after && typeof this.after === 'function') {
      return this.after(value);
    }

    return value;
  }

  doSomething(value) {
    var beforeVal = this._beforeHook(value);
    var fooVal = foo(beforeVal);
    return this._afterHook(fooVal);
  }
}
```

As is, they act as passthroughs with the added benefit that `doSomething` doesn't reference the hooks directly. It's a barrier between the library code and the user's code. The isolation is good because the user's code doesn't interact directly with the internal code and vice versa. Either can change without breaking the other.

But say that we don't want a simple passthrough:

``` js
class ThingWithHooks {
  _beforeHook(value) {
    doSomethingElseBeforeDoSomething();
    if (this.before && typeof this.before === 'function') {
      return this.before(value);
    }

    return value;
  }

  _afterHook(value) {
   doSomethingElseAfterDoSomething()
    if (this.after && typeof this.after === 'function') {
      return this.after(value);
    }

    return value;
  }

  doSomething(value) {
    var beforeVal = this._beforeHook(value);
    var fooVal = foo(beforeVal);
    return this._afterHook(fooVal);
  }
}
```

I added my own before and after logic that gets executed without changing the `doSomething` function and it doesn't affect the user's code either. **The public functions didn't change.**

If you go with this approach, you could move the `checkOk` into the `_afterResponse` method which simplifies the `send` function overall and reduces its responsibility. You could do the same with the config and url resolving if you wanted.

I also think it's a lot cleaner of an implementation.

euwest replied:
> The isolation is good because the user's code doesn't interact directly with the internal code and vice versa.

I'm not sure what you mean by "interact directly", because by my definition it does exactly that, just via a cup-game of function calls.

> I added my own before and after logic that gets executed without changing the doSomething function and it doesn't affect the user's code either.

the problem with your examples is they are function calls that, presumably, have some sort of side effect that is indirectly important is some way. (i.e. you're not using the return value of `doSomethingElseBeforeDoSomething()`) That being the case you can't say that it didn't affect the user's code, it just didn't _noticeably_ affect it. who knows what it did via side effects. (because presumably it's not a no-op).

> I added my own before and after logic that gets executed without changing the doSomething function

that's achievable, and more flexible, with the current implementation and using `super`, because whatever changes i want to make can be in the before and after stub functions, and users that extend to override those functions can choose to `super` in my additions or not, at their discretion.

> It's a subtle difference for stronger code.

I don't see the strength you're referring to. I see extraneous type-checking and masking of functions to keep users from taking advantage of class inheritance, which is part of es2015.

> I also think it's a lot cleaner of an implementation.

I think it's the exact opposite. 😦

LionOps replied:
> the problem with your examples is they are function calls that, presumably, have some sort of side effect that is indirectly important is some way. (i.e. you're not using the return value of doSomethingElseBeforeDoSomething())

I think you missed the point. The functions don't have to have side effects, they can just as easily manipulate the value.

> that's achievable, and more flexible, with the current implementation and using super, because whatever changes i want to make can be in the before and after stub functions, and users that extend to override those functions can choose to super in my additions or not, at their discretion.

Then these aren't hooks. They're class methods intended to be overwritten. Hooks don't necessitate a class extension, and while you can say that people can manipulate the methods directly to do what they want, you can't do that AND get the benefit of changes to stipulate. Right now, there is no reason to use `super` because the functions are literally placeholders. If they're suddenly not placeholders, then users that didn't use `super` are SOL and can't get the benefit of whatever changes you made.

To me, that seems like poor design.

> I don't see the strength you're referring to. I see extraneous type-checking and masking of functions to keep users from taking advantage of class inheritance, which is part of es2015.

There are issues with class inheritance from a design standpoint; not every user is going to want that. I don't see the point in shoehorning them into class inheritance or mucking with the base class to get the functionality you want to give them.

euwest replied:
> Then these aren't hooks. They're class methods intended to be overwritten.

Ok, I'll stop calling them hooks then. I thought it was pretty clear what I meant by using that word, but it sounds like you've got a conflicting definition in your mind, so we can drop the word entirely. I agree with the second sentence of the above quote.

> Right now, there is no reason to use super because the functions are literally placeholders. If they're suddenly not placeholders, then users that didn't use super are SOL and can't get the benefit of whatever changes you made.

this inverse of this (i.e. what you're proposing) is to force users to use the changes, by not making it an opt-in (user bumps the version which means `_beforeHook` now has `doSomethingElseBeforeDoSomething()` in it. instead of being able to `super` or not, the choice they have regarding `doSomethingElseBeforeDoSomething` is "use it or downgrade").

> If they're suddenly not placeholders, then users that didn't use super are SOL and can't get the benefit of whatever changes you made.

You're looking at this backwards. they're not "SOL", it's just not going to change behavior without them actively choosing for it to do so (by adding `super`). That's _the definition_ of backwards compatibility.

> There are issues with class inheritance from a design standpoint; not every user is going to want that.

That's true of every design choice. Are you claiming that the other proposed solution is not making any trade-offs? I reject that premise. Are you saying that your proposal serves more people better? I would like to see _any_ supporting evidence. 

> I don't see the point in shoehorning them into class inheritance or mucking with the base class to get the functionality you want to give them.

Here's an alternative to class extension that works with the current implementation (note: at no point have I ever espoused "mucking with the base class"):

``` js
const stip = new Stipulate();

stip.beforeRequest = (url, options) => {
  // do whatever you want
  return [modifiedUrl, modifiedOptions];
};
```

If you're dead set on being able to pass a function as an option on instantiation, we could include the following in the constructor:

```
class Stipulate {
   constructor(options) {
    if(options.before) {
      this.beforeRequest = options.before;
    }
    if(options.after) {
      this.afterResponse = options.after;
    }
    ...
```

beck replied:
Why do we need hooks for an initial release?

euwest replied:
> Why do we need hooks for an initial release?

We don't. I added them because (as currently implemented) they were easy. Shall I drop them?

beck replied:
It is a feature generating a lot of design discussion that is not needed to move us forward.

Yeah, I suggest drop 'em.

euwest replied:
> It is a feature generating a lot of design discussion that is not needed to move us forward.

actually, now that I think about it, in order to drop them we'd also need to drop `JsonStipulate`, or completely re-implement the module the way you brought up (as just a function that wraps `fetch` with error handling and json extraction)

snitch commented on src/index.js:
```diff
@@ -1 +1,34 @@
+      .then(checkOk)
+      .then(this.afterResponse);
+  }
+};
```
Missing semicolon.

morrisallison commented:
I think the existing API is good, but I want to avoid having to use `new` and making an `options` object.

Three things I think would help a lot:
1. Having a `JsonStipulate` class available out of the box.
2. Adding `get`, `post`, `put`, etc. methods without the `options` argument.
3. Adding a static method that creates an instance of the `Stipulate` with the given `options`.
   - I used the name `make` below as an example.

With that, I could write something like this:

``` javascript
import {JsonStipulate} from 'stipulate';

// [...]

let body = { foo: 1, bar: 2 };

JsonStipulate.make({ body }).post('/some/resource', query);

// [...]

JsonStipulate.make({ okCodes: [404] }).get('/another/resource');
```

Instead of this:

``` javascript
import CustomJsonStipulate from './custom-json-stipulate';

// [...]

let stipulation = new CustomJsonStipulate();

let options = {
 method: 'post',
 body: { foo: 1, bar: 2 }
};

stipulation.send('/some/resource', options, query);

// [...]

let stipulation = new CustomJsonStipulate({ okCodes: [404] });
let options = { method: 'get' };

stipulation.send('/some/resource', options);
```

LionOps commented on src/resolve-url.js:
```diff
@@ -0,0 +1,17 @@
+  const mergedQuery = defaults({}, query, parsedUrl.query);
+  const fullQuery = pickBy(mergedQuery, (value) => value);
+
+  delete parsedUrl.search;
```
Is this actually necessary? I want to say that `url` is smart enough to prefer `query` to search, but I might be wrong.

euwest replied:
> I want to say that url is smart enough to prefer query to search

<img width="321" alt="screen shot 2016-04-13 at 3 09 25 pm" src="https://cloud.githubusercontent.com/assets/6263840/14511057/c71f2a54-0189-11e6-8bd5-951ff43fd9cb.png">

LionOps commented on test/make-ok-check.js:
```diff
@@ -0,0 +1,131 @@
+      expect(checkErrMsg).to.throw(Error, 'Internal Server Error');
+    });
+
+    it('uses "Bad Response" as the error message when not statusText', () => {
```
Should it instead use `Response.error`?

euwest replied:
That function is for network errors:

> The static error() method, when invoked, must return a new Response object, associated with a new network error and a new Headers object whose guard is "immutable".

LionOps commented on test/pre-post-extension.js:
```diff
@@ -0,0 +1,56 @@
+
+chai.use(chaiAsPromised);
+
+class ExtStip extends Stipulate {
```
Change the filename to reflect before/after?

euwest replied:
Do you have a name in mind? I think it's fine as is, but I'll change it if you've got one you want.

LionOps replied:
If you adapt to what I said about hooks, this entire file can be deleted. Especially since these tests are actually testing class inheritance more so than actual functionality.

euwest replied:
I disagree; It's testing that by extending the class in such a way, your functions will be used by send in the expected manner. Do you not think that's worth testing?

LionOps replied:
You're testing the EMCA 6 specification for class extension, which shouldn't be tested. 

In the index test file, you should be asserting that the hooks are called before and after the fetch--basically, that the hooks are called when they're supposed to be called.

Extension is clearly defined and an expected behavior. If the hooks are called in `send`, then it follows that they'll be called in an extended class too.

euwest replied:
Ahhhh, I think I see what you're saying now. These tests are showing that 1) `beforeRequest` is being called before the request is being sent, and 2) `afterResponse` is being called after receiving the response. However, I've implemented the tests by extending the class, and (if I gather correctly) you believe the wording of the tests as well as the title of the file obfuscates what's actually being tested. If that's the case, I agree. But I do think these are valid tests, implemented in an informative way(i.e. you can look at them and see an example of how one might extend the class to use those hooks). How about just `hook-methods.js`, and rewrite the test descriptions to be more agnostic about how the hooks are implemented?

euwest replied:
I think I should respond more specifically to a couple things you mentioned:

> You're testing the EMCA 6 specification for class extension, which shouldn't be tested.

It's not being tested, it is being used during setup of the tests. If you're saying the descriptions of the tests and the name of the file are confusing, I now understand why and agree.

> Extension is clearly defined and an expected behavior.

Definitely. I whole-heartedly agree. Which is why [your suggestion](https://github.com/yola/stipulate/pull/2#discussion_r59650144) to eschew its use (i.e. class inheritance) in favor of `_beforeRequest` which looks for `this.before` (or whatever else it ends up being named) doesn't make sense to me. It screams of over-engineering to me when there's a perfectly simple, safe, intelligible, native mechanism to use.

euwest commented:
> Having a `JsonStipulate` class available out of the box.

yep, i'm in to that.

> Adding `get`, `post`, `put`, etc. methods without the `options` argument.

That seems reasonable, except for dropping `options`. Why the restriction?

> Adding a static method that creates an instance of the `Stipulate` with the given options.

Two things regarding this point (and the earlier preface to it in the first paragraph): 
1. I'm not sure I understand the desire to avoid `new`. The library as a whole is based on the idea of creating a container of configuration for `fetch`, from which you can make successive requests. Hiding `new` in a static method feels like trying to hide one of the main features of the module.
2. Your example with `make` is essentially stretching the fetch API out into a chain of method calls--much like [`superagent`](https://github.com/visionmedia/superagent)--and is actively negating one of the main functions of the library: configuration persistence across multiple requests. Am I misunderstanding what you're asking for?

We could create a static method that allows you to fire off requests without instantiating the class. I'm still not sure we want to mimic superagent's method chaining, but if that's the direction the group wants to go I won't fight it.

morrisallison commented:
> Am I misunderstanding what you're asking for?

I'm asking for more, and simpler ways to send the requests. I'm not suggesting we always recreate the configuration for every request.

> I'm not sure I understand the desire to avoid `new`.

Avoiding `new` is mainly a API design preference, I could just as easily write `new Stipulate().send(url)`.

> That seems reasonable, except for dropping `options`. Why the restriction?

If the request needs to provide additional options, then it could just use `send` and provide the `method` in those options.

I don't see much point in doing this:

``` javascript
const fetchResource = function(options, query) {
  const stipulateOptions = {
    headers: { 'Cache-Control': 'no-cache' }
  };

  stipulation.get('/some/resource', stipulateOptions, query);
};
```

Over this:

``` javascript
const fetchResource = function(options, query) {
  const stipulateOptions = {
    method: 'GET',
    headers: { 'Cache-Control': 'no-cache' }
  };

  return stipulation.send('/some/resource', stipulateOptions, query);
};
```

but in the case where only the `method` option is needed, I'd like to express only that.

``` javascript
const fetchResource = function(options, query) {
  return stipulation.get('/some/resource', query);
};
```

I'm suggesting that the expressive methods serve a specific purpose beyond just assigning the `method` option.

For example, @DevShep was talking about a `body` argument before. I'd like to see `.put()` and `.post()` have that argument.

``` javascript
const fetchResource = function(options, query) {
  const body = {
    foo: options.foo,
    bar: new Date(options.bar).toString();
  };

  return stipulation.post('/some/resource', body, query);
};
```

_edit:_

Side note: I'm not too hung up on these ideas. In the end, keeping everything consistent, may be better.

This isn't bad either:

``` javascript
stipulation.post('/some/resource', {body}, query);
```

euwest commented:
> Side note: I'm not too hung up on these ideas. In the end, keeping everything consistent, may be better.

I think this is a valuable thing to hash out, but maybe in terms of future releases. I think I agree with you on keeping everything consistent (and as a consequence, simple) for now but let's keep talking about this*.

*maybe open a ticket with the request for http verb methods, and link to this comment thread

euwest committed:
add JsonStipulate to exports, tests for class

snitch commented on src/index.js:
```diff
@@ -1 +1,62 @@
+
+    return fetch(...requestArguments)
+      .then(checkOk)
+      .then(this.afterResponse.bind(this));
```
Unnecessary semicolon.

snitch commented on test/json-stipulate.js:
```diff
@@ -0,0 +1,76 @@
+        }
+      };
+      const url = '/boop';
+      const expected = "{\"foo\":true}";
```
Mixed double and single quotes.

euwest committed:
fix mixed string linting error

snitch commented:
:sparkles: No lint errors found. :sparkles:

euwest committed:
rename hook method test file, reword tests

file name and test descriptions were unclear/misleading about what they were testing.

snitch commented:
:sparkles: No lint errors found. :sparkles:

beck commented:
I'm surprise to see classes and having to create an instance of anything to use this helper.

Why not write `stipulate` using the same API as `fetch`?

With the following behavior:
- throws if not okay
- promise resolves to the data one _stipulates_ must be there

``` js
import stipulate from 'stipulate';
let resourceA = stipulate('/serve/some/resource');
let resourceB = stipulate('/serve/some/resource', { okCodes: [401, 403] });
stipulate('/service/action', {method: 'POST'});
```

euwest committed:
rework stipulate to simplify, and remove class structure

instead of encapsulating all of stipulates functionality in a class, just export each piece
and let consumers pick and choose what they use

snitch commented:
:sparkles: No lint errors found. :sparkles:

beck commented on README.md:
````diff
@@ -1,6 +1,137 @@
+```js
+import stipulate from 'stipulate';
+
+const options = {
````
I would expect the first examples to be the most simple use case without options.  Something like:

``` js
stipulate('http://www.mocky.io/v2/5710408d110000100a9e267a')
  .then(console.log);

// console:
// { connor: 'rocks' }


stipulate('http://www.mocky.io/v2/571040f0110000430a9e267b')
  .catch(e => console.log('Fail: ' + e.response.status));

// console:
// Fail: 404
```

beck commented on README.md:
```diff
@@ -1,6 +1,137 @@
+  }
+};
+
+const responsePromise = stipulate('/some/endpoint', options);
```
Minor:

``` js
const resourcePromise = stipulate('/service/resource', options);
```

Reenforces `stipulate` abstracts away need to worry about responses

beck commented on README.md:
```diff
@@ -1,6 +1,137 @@
+
+### Errors
+
+By default, Stipulate throws responses with non 2XX status codes. There are two ways to modify this
```
Since we're using promises, is it better to say "Stipulate rejects".

Otherwise it sounds like you'd need to wrap stipulate calls in a try/catch.

beck commented on test/build-options.js:
```diff
@@ -0,0 +1,79 @@
+    const a = {
+      foo: 'bar',
+      fizz: {
+        buzz: 'zip'
```
can you use values that are easier to distinguish?
something like:

```
a.fizz.buzz = 'declared first - GOOD';
b.fizz.buzz = 'overwrite - BAD'; 
```

beck commented on test/index.js:
```diff
@@ -1,6 +1,46 @@
-import {expect} from 'chai';
```
Should this file name be `test/speculate.js` or `test/speculate-spec.js`?

beck commented on test/index.js:
```diff
@@ -1,6 +1,46 @@
+// mock response that can respond to different data
+// extraction methods.
+const responseMock = (url, options) => {
+  const jsonData = Promise.resolve([url, options]);
```
This is slightly confusing. Can you make this some dummy value that can be asserted in the test?

My confusion:

```
const results = stipulate(url, opts);
return expect(results).to.eventually.deep.equal([url, opts]);
```

Makes me think that stipulate is supposed to return what you pass in.

Instead, if it was:

```
const resource = stipulate(url, opts);
return expect(resource).to.eventually.deep.equal({some:'dummy data'});
```

then this would be perfectly clear.

beck commented on test/make-ok-check.js:
```diff
@@ -0,0 +1,131 @@
+import enforceOk from '../src/enforce-ok';
```
rename this test file to match? `test/enforce-ok.js`?

beck commented on test/build-options.js:
```diff
@@ -0,0 +1,79 @@
+    expect(result).to.deep.equal({ foo: 'bar', fizz: 'buzz' });
+  });
+
+  it('doesnt have to be given anything', () => {
```
This test name doesn't exactly describe what is under review. There is an assertion being made on the response.

beck commented:
Most of my critique is about language.

Overall this LGTM.

euwest committed:
update readme language

euwest committed:
update language, clarify test case variable names/values, fix test file names

snitch commented:
:sparkles: No lint errors found. :sparkles:

euwest committed:
alphabetize imports

snitch commented:
:sparkles: No lint errors found. :sparkles:

LionOps commented on README.md:
```diff
@@ -1,6 +1,150 @@
+
+By default, after checking for errors Stipulate will try to extract json from the response to return in a promise.
+If you want another data type (text and blob are some examples of other supported response data types) you can
+pass a third argument to stipulate, which is the data type's name that you want.
```
I would either enumerate or link to fetch documentation to provide what those types are.

LionOps commented:
👍  Other than my nitpick.

euwest committed:
add link to body specs in readme for data extraction section

snitch commented:
:sparkles: No lint errors found. :sparkles:

euwest committed:
ready 0.0.1

snitch commented:
:sparkles: No lint errors found. :sparkles:

beck commented:
Version bumps should happen in a separate pull.

:+1:

euwest commented:
> Version bumps should happen in a separate pull.

it's not a bump, it's the initial version
